File: fb/24.Swap-Nodes-in-Pairs.cpp
// Time O(n)
// Space O(1)
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode dummy(0,head);
        ListNode* pre = &dummy;
        while(pre->next && pre->next->next){
            ListNode* first = pre->next;
            ListNode* second = first->next;
            ListNode* third = second->next;
            pre->next = second;
            first->next = third;
            second->next = first;
            // next pair
            pre = first;  // pre point to actually the second
        }
        return dummy.next;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

File: fb/24.Swap-Nodes-in-Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fakeHead = ListNode(0, head)
        pre = fakeHead
        
        while pre.next and pre.next.next:
            first = pre
            second = pre.next
            third = pre.next.next
            
            second.next = third.next
            third.next = second
            first.next = third
            
            pre = second
        return fakeHead.next
